crates/pera-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -181,155 +332,270 @@ impl DWalletMPCManager {
+                            .party_to_weight
+                            .len();
+
+                // The manager must hold a Network key to advance.
+                // The only exception is if this is the DKG session
+                // that creates and initializes this key for the first time.
+                let is_manager_ready = !cfg!(feature = "with-network-dkg")
+                    || (is_valid_network_dkg_transaction
                         || matches!(
                             mpc_network_key_status,
+                            // Todo (yael): Check if the current relevant key version exist
                             DwalletMPCNetworkKeysStatus::Ready(_)
-                        )
-                }
-                if is_ready && is_manager_ready {
-                    Some(session)
+                        ));
+                if session.is_ready_to_advance() && is_manager_ready {
+                    // We must first clone the session, as we approve to advance the current session
+                    // in the current round and then start waiting for the next round's messages
+                    // until it is ready to advance or finalized.
+                    let session_clone = session.clone();
+                    session.pending_quorum_for_highest_round_number =
+                        session.pending_quorum_for_highest_round_number + 1;
+                    Some(session_clone)
                 } else {
                     None
                 }
             })
-            .collect::<Vec<&mut DWalletMPCSession>>();
-
-        let mut malicious_parties = vec![];
-        let mut messages = vec![];
-        ready_to_advance
-            .par_iter_mut()
-            .map(|session| {
-                (
-                    session.advance(&self.weighted_threshold_access_structure, self.party_id),
-                    session.session_info.session_id,
-                )
-            })
-            .collect::<Vec<_>>()
-            // Convert back to an iterator for processing.
-            .into_iter()
-            .try_for_each(|(result, session_id)| match result {
-                Ok((message, malicious)) => {
-                    messages.push((message, session_id));
-                    malicious_parties.extend(malicious);
-                    Ok(())
+            .collect();
+
+        self.cryptographic_computations_orchestrator
+            .insert_ready_sessions(sessions_ready_to_advance);
+        Ok(())
+    }
+
+    /// Spawns all ready MPC cryptographic computations using Rayon.
+    /// If no local CPUs are available, computations will execute as CPUs are freed.
+    pub(crate) fn perform_cryptographic_computation(&mut self) {
+        while self
+            .cryptographic_computations_orchestrator
+            .currently_running_sessions_count
+            < self
+                .cryptographic_computations_orchestrator
+                .available_cores_for_cryptographic_computations
```

Comment:
Couldn't this be written as

```suggestion
        while self
            .cryptographic_computations_orchestrator
            .has_available_core()
```